pkg/image/containerd/daemon_provider.go

```diff
@@ -0,0 +1,485 @@
+	}
+
+	return &DaemonImageProvider{
+		imageStr:  imgStr,
+		tmpDirGen: tmpDirGen,
+		client:    c,
+		platform:  platform,
+		namespace: namespace,
+	}, nil
+}
+
+// Provide an image object that represents the cached docker image tar fetched from a containerd daemon.
+func (p *DaemonImageProvider) Provide(ctx context.Context, userMetadata ...image.AdditionalMetadata) (*image.Image, error) {
+	ctx = namespaces.WithNamespace(ctx, p.namespace)
+
+	resolvedImage, resolvedPlatform, err := p.pullImageIfMissing(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tarFileName, err := p.saveImage(ctx, resolvedImage)
+	if err != nil {
+		return nil, err
+	}
+
+	// use the existing tarball provider to process what was pulled from the containerd daemon
+	return sdocker.NewProviderFromTarball(tarFileName, p.tmpDirGen).
+		Provide(ctx,
+			withMetadata(resolvedPlatform, userMetadata, p.imageStr)...,
+		)
+}
+
+// pull a containerd image
+func (p *DaemonImageProvider) pull(ctx context.Context, resolvedImage string) (containerd.Image, error) {
+	var platformStr string
+	if p.platform != nil {
+		platformStr = p.platform.String()
+	}
+
+	// note: if not platform is provided then containerd will default to linux/amd64 automatically. We don't override
+	// this behavior here and intentionally show that the value is blank in the log.
+	log.WithFields("image", resolvedImage, "platform", platformStr).Debug("pulling containerd")
+
+	ongoing := newJobs(resolvedImage)
+
+	// publish a pull event on the bus, allowing for read-only consumption of status
+	bus.Publish(partybus.Event{
+		Type:   event.PullContainerdImage,
+		Source: resolvedImage,
+		Value:  newPullStatus(p.client, ongoing).start(ctx),
+	})
+
+	h := images.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		// as new layers (and other artifacts) are discovered, add them to the ongoing list of things to monitor while pulling
+		if desc.MediaType != images.MediaTypeDockerSchema1Manifest {
+			ongoing.Add(desc)
+		}
+		return nil, nil
+	})
+
+	options := p.pullOptions(ctx)
+	options = append(options, containerd.WithImageHandler(h))
+	if platformStr != "" {
+		options = append(options, containerd.WithPlatform(platformStr))
+	}
+
+	// note: this will return an image object with the platform correctly set (if it exists)
+	resp, err := p.client.Pull(ctx, resolvedImage, options...)
+	if err != nil {
+		return nil, fmt.Errorf("pull failed: %w", err)
+	}
+
+	return resp, nil
+}
+
+func (p *DaemonImageProvider) pullOptions(ctx context.Context) []containerd.RemoteOpt {
+	var options = []containerd.RemoteOpt{
+		containerd.WithPlatform(p.platform.String()),
+	}
+
+	doptions := docker.ResolverOptions{
+		Tracker: docker.NewInMemoryTracker(),
+	}
+
+	useRegAuth, err := strconv.ParseBool(os.Getenv("USE_REGISTRY_AUTH"))
+	if err != nil {
+		useRegAuth = false
+	}
+
+	if useRegAuth {
+		username := os.Getenv("CTR_REGISTRY_USERNAME")
+		secret := os.Getenv("CTR_REGISTRY_PASSWORD")
```

Comment:
the CONTAINERD_ADDRESS is, I'll double check on the rest

Reply:
no these don't look standard -- @shanedell are these env var names used in other applications? Or were they made in this PR as a necessity?

Reply:
I haven't been able to find anything yet. It seems that the registry options map quite nicely onto the containerd options at first glance. I'll look into mapping these options and requiring that input for containerd API access.

Reply:
@wagoodman I can't currently remember where I got these environment variables from. I believe it should have from either the [containerd](https://github.com/containerd/containerd) repo or the [nerdctl](https://github.com/containerd/nerdctl) repo as I looked at these both extensively to figure out how things were working. I may have just made some of those up on the fly at the time to make sure the authentication could work and can be updated.

Reply:
I wasn't able to reproduce using these env vars with nerdctl or ctr commands (image pull, login, etc). A quick search in each repo also didn't really show any equivalent env vars. I'm going to go the path of mapping this onto the existing `RegistryOptions`, since containerd is a more low-level tool it seems to require a lot of these options force fed as CLI options directly... which hints that it's really acting more like the OCI registry provider anyway.

I've got a mock up locally, I'll get it cleaned up and pushed tonight (maybe)